pkg/querier/block_streaming.go

```diff
@@ -0,0 +1,280 @@
+type blockStreamingQuerierSeriesSet struct {
+	series       []*storepb.StreamingSeries
+	streamReader chunkStreamReader
+
+	// next response to process
+	nextSeriesIndex int
+
+	currSeries storage.Series
+}
+
+type chunkStreamReader interface {
+	GetChunks(seriesIndex uint64) ([]storepb.AggrChunk, error)
+}
+
+func (bqss *blockStreamingQuerierSeriesSet) Next() bool {
+	bqss.currSeries = nil
+
+	if bqss.nextSeriesIndex >= len(bqss.series) {
+		return false
+	}
+
+	currLabels := mimirpb.FromLabelAdaptersToLabels(bqss.series[bqss.nextSeriesIndex].Labels)
+	seriesIdxStart := bqss.nextSeriesIndex // First series in this group. We might merge with more below.
+	bqss.nextSeriesIndex++
+
+	// Chunks may come in multiple responses, but as soon as the response has chunks for a new series,
+	// we can stop searching. Series are sorted. See documentation for StoreClient.Series call for details.
+	// The actually merging of chunks happens in the Iterator() call where chunks are fetched.
+	for bqss.nextSeriesIndex < len(bqss.series) && labels.Compare(currLabels, mimirpb.FromLabelAdaptersToLabels(bqss.series[bqss.nextSeriesIndex].Labels)) == 0 {
+		bqss.nextSeriesIndex++
+	}
+
+	bqss.currSeries = newBlockStreamingQuerierSeries(currLabels, seriesIdxStart, bqss.nextSeriesIndex-1, bqss.streamReader)
+	return true
+}
+
+func (bqss *blockStreamingQuerierSeriesSet) At() storage.Series {
+	return bqss.currSeries
+}
+
+func (bqss *blockStreamingQuerierSeriesSet) Err() error {
+	return nil
+}
+
+func (bqss *blockStreamingQuerierSeriesSet) Warnings() storage.Warnings {
+	return nil
+}
+
+// newBlockStreamingQuerierSeries makes a new blockQuerierSeries. Input labels must be already sorted by name.
+func newBlockStreamingQuerierSeries(lbls labels.Labels, seriesIdxStart, seriesIdxEnd int, streamReader chunkStreamReader) *blockStreamingQuerierSeries {
+	return &blockStreamingQuerierSeries{
+		labels:         lbls,
+		seriesIdxStart: seriesIdxStart,
+		seriesIdxEnd:   seriesIdxEnd,
+		streamReader:   streamReader,
+	}
+}
+
+type blockStreamingQuerierSeries struct {
+	labels                       labels.Labels
+	seriesIdxStart, seriesIdxEnd int
+	streamReader                 chunkStreamReader
+}
+
+func (bqs *blockStreamingQuerierSeries) Labels() labels.Labels {
+	return bqs.labels
+}
+
+func (bqs *blockStreamingQuerierSeries) Iterator(reuse chunkenc.Iterator) chunkenc.Iterator {
+	// Fetch the chunks from the stream.
+	var allChunks []storepb.AggrChunk
+	for i := bqs.seriesIdxStart; i <= bqs.seriesIdxEnd; i++ {
+		chks, err := bqs.streamReader.GetChunks(uint64(i))
+		if err != nil {
+			return series.NewErrIterator(err)
+		}
+		allChunks = append(allChunks, chks...)
+	}
+	if len(allChunks) == 0 {
+		// should not happen in practice, but we have a unit test for it
+		return series.NewErrIterator(errors.New("no chunks"))
+	}
+
+	sort.Slice(allChunks, func(i, j int) bool {
+		return allChunks[i].MinTime < allChunks[j].MinTime
+	})
```

Comment:
Are chunks not already sorted by the store-gateway?

Reply:
@dimitarvdimitrov are they always sorted even when multiple blocks are involved?

Reply:
They aren't necessarily. They will be in most cases except when there are overlapping blocks. In those cases the order will be: first chunks from one block (ordered by minTime), then chunks from the second block (ordered by minTime).

Reply:
this is the code which merges chunks ranges from different blocks https://github.com/grafana/mimir/blob/69acff88dda466f7150397f1be7999864f13138d/pkg/storegateway/series_refs.go#L508-L532